src/lib/operators/join_nested_loop.cpp

```diff
@@ -0,0 +1,264 @@
+
+                if (is_outer_join) {
+                  left_matches[left_value.chunk_offset()] = true;
+                }
+
+                if (_mode == JoinMode::Outer) {
+                  right_matches.insert(RowID{chunk_id_right, right_value.chunk_offset()});
+                }
+              }
+            });
+          });
+        });
+      }
+
+      if (is_outer_join) {
+        // add unmatched rows for outer joins
+        for (ChunkOffset chunk_offset{0}; chunk_offset < left_matches.size(); ++chunk_offset) {
+          if (!left_matches[chunk_offset]) {
+            pos_list_left->emplace_back(RowID{chunk_id_left, chunk_offset});
+            pos_list_right->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+          }
+        }
+      }
+    });
+  }
+
+  // For Full Outer we need to add all unmatched rows for the right side.
+  // Unmatched rows on the left side are already added in the main loop above
+  if (_mode == JoinMode::Outer) {
+    for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < right_table->chunk_count(); ++chunk_id_right) {
+      auto column_right = right_table->get_chunk(chunk_id_right).get_column(right_column_id);
+
+      resolve_column_type<RightType>(*column_right, [&](auto& typed_right_column) {
+        auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+        iterable_right.for_each([&](const auto& right_value) {
+          const auto row_id = RowID{chunk_id_right, right_value.chunk_offset()};
+          if (!right_matches.count(row_id)) {
+            pos_list_left->emplace_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+            pos_list_right->emplace_back(row_id);
+          }
+        });
+      });
+    }
+  }
+
+  // write output chunks
+  auto output_chunk = Chunk();
+
+  if (_mode == JoinMode::Right) {
+    _write_output_chunks(output_chunk, right_table, pos_list_right);
+    _write_output_chunks(output_chunk, left_table, pos_list_left);
+  } else {
+    _write_output_chunks(output_chunk, left_table, pos_list_left);
+    _write_output_chunks(output_chunk, right_table, pos_list_right);
+  }
+
+  _output_table->emplace_chunk(std::move(output_chunk));
+}
+
+void JoinNestedLoop::_write_output_chunks(Chunk& output_chunk, const std::shared_ptr<const Table> input_table,
+                                          std::shared_ptr<PosList> pos_list) {
+  // Add columns from table to output chunk
+  for (ColumnID column_id{0}; column_id < input_table->column_count(); ++column_id) {
+    std::shared_ptr<BaseColumn> column;
+
+    if (auto ref_col_left = std::dynamic_pointer_cast<const ReferenceColumn>(
+            input_table->get_chunk(ChunkID{0}).get_column(column_id))) {
+      auto new_pos_list = std::make_shared<PosList>();
+
+      ChunkID current_chunk_id{0};
+
+      // de-reference to the correct RowID so the output can be used in a Multi Join
+      for (const auto& row : *pos_list) {
```

Comment:
`row` is an integer, isn't it? If yes, we shouldn't take it by reference.

Reply:
it's a `RowID`, so a `struct` and not a primitive.

Reply:
A 64-bit one though. My intuition is that we don't need the indirection here. I don't think we have strict rules on this, do we?

Reply:
The compiler is probably going to take care of this, right?

Reply:
good point, changed it